Readme.md
| Note: | This client library has been deprecated in favor of http://code.google.com/p/google-api-ads-ruby |
|:------|:-------------------------------------------------------------------------------------------------|


---


# Google adwords4r Library #

Welcome to adwords4r: Bringing the delights of Ruby programming to the AdWords world!

Documentation and comments are a work in progress.


# Useful Web Resources #

  * [AdWords home](https://adwords.google.com/)
  * [AdWords API discussion forum](http://groups.google.com/group/adwords-api)
  * [AdWords API documentation](http://code.google.com/apis/adwords/docs/developer/index.html)
  * [This project's Google Code page](http://code.google.com/p/google-api-adwords-ruby/)


# Docs for Users #

## 1 - Installation: ##

adwords4r is a ruby gem.  See http://docs.rubygems.org/read/book/1

Install it using the gem install command.
```
$ gem install --remote adwords4r
```

The following gem libraries are required:
  * soap4r v1.5.8
  * httpclient v2.1.2 or greater


## 2 - Using the client library: ##

It's pretty easy to use.
See http://docs.rubygems.org/read/chapter/3#page70 for how to set the rubygem
environment.
```
$ export RUBYOPT=rubygems
```
or
```
$ ruby -rubygems my_program_that_uses_gems
```

If you do not use the rubygems option, you need to add
```
require 'rubygems'
```
at the beginning of your programs.

Then:
```
require 'adwords4r'

adwords = AdWords::API.new
```
creates a driver for the latest version of AdWords API using a credentials file
in ENV['HOME']/adwords.properties
There is an example credentials in the root adwords4r directory.
You can also pass API a manually constructed AdWordsCredentials object like:
```
adwords = AdWords::API.new(AdWords::AdWordsCredentials.new({
  'developerToken' => 'DEVELOPER_TOKEN',
  'applicationToken' => 'APPLICATION_TOKEN',
  'useragent' => 'Ruby Sample',
  'password' => 'PASSWORD',
  'email' => 'user@domain.com'
  'clientEmail' => 'user2@domain.com',
  'environment' => 'PRODUCTION',
}))
```

Then, just specify which service you're looking to use, and which version:
```
account_srv = adwords.get_service(13, 'Account')
```

and you should now be able to just use the API methods in the object you were
returned:
```
client_accounts = account_srv.getClientAccounts
```

See the code in the examples directory for working examples you can build from.

**Note**: if your setup requires you to send connections through a proxy server,
please set the HTTP\_PROXY environment variable. For example, from your code:
```
$ ENV['HTTP_PROXY'] = 'http://myproxyserver:8080'
```


### 2.1 - Using the Sandbox and other environments: ###

In order to use the v13 and v2009 sandboxes, make sure that the 'environment'
parameter in the credentials is set to SANDBOX.
```
environment=SANDBOX
```

Should you need to use an environment other than production or sandbox, you can
add it to the list of environments in runtime:
```
AdWords::Service::add_environment('ENVIRONMENT_NAME', {
    13 => 'URL_FOR_v13',
    200906 => 'URL_FOR_v200906'
})
```
and then use it normally in your code:
```
credentials.environment = 'ENVIRONMENT_NAME'
```

Make sure you only include the base URL, that is, the URL up until the "adwords"
part, leaving the version and group out. Here's an example:
```
AdWords::Service::add_environment('SANDBOX', {
    13 => 'https://sandbox.google.com/api/adwords/',
    200906 => 'https://adwords-sandbox.google.com/api/adwords/'
}
```


### 2.2 - Mixing v13 and v2009: ###

It's possible to access both the v13 and v2009 sandboxes or production
environments simultaneously. Since both versions share a common backend storage,
it is possible for your application to choose between v13 and v2009 services
at will, mixing and matching them.

The multiple\_versions.rb code sample shows you how to build an application using
v13 and v2009 services simultaneously.


### 2.3 - Logging: ###

It is often useful to see a trace of the raw SOAP XML being sent and received.
The quickest way of achieving this when debugging your application is by setting
the ADWORDS4R\_DEBUG environment variable to TRUE; e.g. in the bash shell:
```
$ export ADWORDS4R_DEBUG=TRUE
```
or from your Ruby code:
```
ENV['ADWORDS4R_DEBUG'] = 'TRUE'
```

This will output the SOAP XML to stderr, which will usually show up in your
terminal window.

There's also the option of logging requests and XML to files. In order to enable
this, you should use the write\_to\_file method of the loggers inside your AdWords::API object:
```
adwords = AdWords::API.new
adwords.unit_logger.log_to_file
adwords.xml_logger.log_to_file
```

The first logs the units spent and the requests made, whereas the second logs
the full SOAP XML being sent and received over the wire. These will be written
to the current directory by default, but you can specify a path as an optional
parameter to log\_to\_file:
```
adwords.unit_logger.log_to_file('/var/log/my_app')
adwords.xml_logger.log_to_file('/var/log/my_app')
```

The files will be named request\_info and soap\_xml, respectively.


# Docs for Developers #

```
$ rake getwsdl
```
to get the wsdl files

```
$ rake generate
```
to regenerate the bindings if needed

```
$ rake rdoc
```
to regenerate the project documentation if needed

```
$ rake package
```
to package the gem and create a release

```
$ rake test
```
to run unit tests on the library


# Copyright/License Info #

Copyright 2009, Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

> http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

ChangeLog.txt
19.4.0
- Added support for the returnMoneyInMicros header.
- Added handle_partial_failures.rb example for v201008.

19.3.0
- Added support for CustomerSyncService, ServicedAccountService and
  UserListService in v201008.
- Added several examples for new services.
- Updated get_conversion_optimizer_eligibility.rb to new fields in v201008.
- Removed CSV files and utility methods; they've been deprecated by the
  documentation.

19.2.0
- Added support for v201008.
- Added several examples for v201008.
- Added utility to help with error handling (operation_index_for_error)

19.1.0
- Added support for v201003.
- Added several examples for v201003, removed deprecated examples.
- Reorganized service tests into versions.
- Added extension methods for v201003 report downloading.

19.0.0
- Removed support for v13 and v13-dependent utilities.

18.0.0
- Removed support for v200906, which has been deco'ed.
- Made the application token optional, since it's been deco'ed.

17.3.0
- Added functionality tests for v2009 CampaignService and AdGroupService.

17.2.0
- Added new set of code samples for v200909.

17.1.0
- Added support for AdParamService in v200909.
- Improved request logging, by adding isFault and faultMessage.
  isFault indicates whether the request was a fault, and faultMessage contains
  the error message, if any.

17.0.1
- Fixed unit tracking for failed requests.
  Issue 13, http://code.google.com/p/google-api-adwords-ruby/issues/detail?id=13

17.0.0
- Added support for v200909.
- Changed the parameter order in API.get_service (aka API.service) to be
  get_service(name, version). The old ordering is still supported but
  deprecated. The version parameter is now optional and defaults to 13.
- Added 'module' method to all the services.
  This returns a reference to the module containing all the classes for the
  service, so the full class name isn't necessary. For example,
  service_object.module::ClassName instead of
  AdWords::V13::SomeService::ClassName. Using this will also make it easier to
  upgrade between API versions, since it removes the need for hardcoded numbers.
- Added 'use_mcc' property to the API object to make it easier to make requests
  at the MCC level. Setting it to true will make all requests be sent with no
  clientEmail or clientCustomerId. It's also possible to invoke the 'use_mcc'
  method with a block, and only that block will be executed at the MCC level.
- Added 'validateOnly' property to the API object to make it easier to send
  requests with the 'validateOnly' header in v2009. This header makes it
  possible to check a request for validity without executing it.
- Changes in handling 'alternateUrl'; now it only errors out if there's no
  'environment' defined.
- CSV data updated for operation rates, now has an extra column at the beginning
  for API version.
- Added several tests
- Moved wrapper generation to its own file, generator.rb. This cleans up
  services.rb.
- Added logging of the "operator" for v2009 mutate requests.
- Fixed sample adwords.properties to use 'environment'.
- Fixed problem with empty lines in adwords.properties.
- Fixed problem making Production the default environment.

16.0.0
- Removed support for the v200902 Sandbox, which has been deco'ed.
- Added support for the v200906 Sandbox and Production environments.
- Changed handling of service environments in the credentials.
  The 'alternateUrl' parameter has been replaced with the 'environment' one.
  This can be set to either PRODUCTION or SANDBOX. It controls which environment
  both v13 and v200906 connect to.

15.1.0
- Added downloadCsvReport extension to the Report Service, downloads a report as
  CSV data.
- Added getMethodUsage extension to the Info Service, returns a breakdown of API
  usage per method.
- Added getClientUnitsUsage extension to the Info Service, returns a smart
  breakdown of API usage per account.
- Service wrappers are now generated during the "rake generate" step of setup,
  instead of at runtime.
  The files are created in the lib/adwords4r/v* directories.
- Added client-side type and arity safety checks for API calls.
  An ArgumentError will now be thrown if an API call gets an invalid number of
  arguments or arguments of an incorrect type.
- Added CSV files with various lists of information (categories, countries,
  etc.) and utility methods to retrieve them. They can be found in the
  newly-created AdWords::Utils module.
- Added several subclasses of ApiError to allow for finer exception handling.
  The correct subclass gets instanced automatically whenever a SOAP fault
  occurs.
- Added unit tests for the client library functionality.
- Added RDoc comments to client library and an RDoc generation task to the
  Rakefile. Pre-generated documentation is also included in the doc/ directory.
- Added unit_spend example and extended the reports one.
- Fixed packaging issues.
- Fixed bug in multiple_versions example.
- Several additions and changes to the Readme.
- Cleaned up Rakefile.

15.0.1
- Bugfix: added generation of ApiError exceptions upon SOAP faults.

15.0.0
- Added support for the v200902 sandbox.
- Added support for multiple API versions on the same AdWords::API object.
- Added request and SOAP XML file logging options.
  This complements the previously existing SOAP XML console output.
- Moved API classes to their own version and service-dependent modules.
  This is necessary for multiple version support.
- Moved API methods from the AdWords::API object to service wrappers.
  This is necessary to avoid collisions between similarly named methods in
  different services.
- Added get_service method to AdWords::API to be able to access a particular
  version of a service.
  It returns the service wrapper containing the API methods for that service.
- Added extensions mechanism, allowing for plugging of client-side methods into
  the wrapper objects.
  This allows us to provide smart methods at the service level that simplify
  common tasks.

trunk/lib/adwords4r/services.rb
#!/usr/bin/ruby
#
# Authors:: sgomes@google.com (Sérgio Gomes)
#           jeffy@google.com (Jeffrey Posnick)
#           chanezon@google.com (Patrick Chanezon)
#           leavengood@gmail.com (Ryan Leavengood)
#
# Copyright:: Copyright 2010, Google Inc. All Rights Reserved.
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# Helper methods for loading and managing the available services.

module AdWords

  # Contains helper methods for loading and managing the available services.
  module Service

    # Set defaults
    DEFAULT_VERSION = 201008
    DEFAULT_ENVIRONMENT = 'PRODUCTION'

    # Configure the services available to each version
    @@services = {
      13 => ['Account', 'Report', 'TrafficEstimator'],
      200909 => ['AdExtensionOverride', 'AdGroupAd', 'AdGroupCriterion',
                 'AdGroup', 'AdParam', 'BulkMutateJob', 'CampaignAdExtension',
                 'CampaignCriterion', 'Campaign', 'CampaignTarget',
                 'GeoLocation', 'Info', 'TargetingIdea'],
      201003 => ['AdExtensionOverride', 'AdGroupAd', 'AdGroupCriterion',
                 'AdGroup', 'AdParam', 'BidLandscape', 'BulkMutateJob',
                 'CampaignAdExtension', 'CampaignCriterion', 'Campaign',
                 'CampaignTarget', 'GeoLocation', 'Info', 'Media',
                 'ReportDefinition', 'TargetingIdea'],
      201008 => ['AdExtensionOverride', 'AdGroupAd', 'AdGroupCriterion',
                 'AdGroup', 'AdParam', 'Alert', 'BidLandscape', 'BulkMutateJob',
                 'CampaignAdExtension', 'CampaignCriterion', 'Campaign',
                 'CampaignTarget', 'CustomerSync', 'Experiment', 'GeoLocation',
                 'Info', 'Media', 'ServicedAccount', 'ReportDefinition',
                 'TargetingIdea', 'TrafficEstimator', 'UserList']
    }

    # Configure the different environments, with the base URL for each one
    @@environments = {
      'PRODUCTION' => {
        13 => 'https://adwords.google.com/api/adwords/',
        200909 => 'https://adwords.google.com/api/adwords/',
        201003 => 'https://adwords.google.com/api/adwords/',
        201008 => 'https://adwords.google.com/api/adwords/',
      },
      'SANDBOX' => {
        13 => 'https://sandbox.google.com/api/adwords/',
        200909 => 'https://adwords-sandbox.google.com/api/adwords/',
        201003 => 'https://adwords-sandbox.google.com/api/adwords/',
        201008 => 'https://adwords-sandbox.google.com/api/adwords/'
      }
    }

    # Configure the subdirectories for each version / service pair.
    # A missing pair means that only the base URL is used.
    @@subdirs = {
      # v13
      [13, 'Account'] => 'v13/',
      [13, 'Report'] => 'v13/',
      [13, 'TrafficEstimator'] => 'v13/',
      # v200909
      [200909, 'AdExtensionOverride'] => 'cm/v200909/',
      [200909, 'AdGroupAd'] => 'cm/v200909/',
      [200909, 'AdGroupCriterion'] => 'cm/v200909/',
      [200909, 'AdGroup'] => 'cm/v200909/',
      [200909, 'AdParam'] => 'cm/v200909/',
      [200909, 'BulkMutateJob'] => 'job/v200909/',
      [200909, 'CampaignAdExtension'] => 'cm/v200909/',
      [200909, 'CampaignCriterion'] => 'cm/v200909/',
      [200909, 'Campaign'] => 'cm/v200909/',
      [200909, 'CampaignTarget'] => 'cm/v200909/',
      [200909, 'GeoLocation'] => 'cm/v200909/',
      [200909, 'Info'] => 'info/v200909/',
      [200909, 'TargetingIdea'] => 'o/v200909/',
      # v201003
      [201003, 'AdExtensionOverride'] => 'cm/v201003/',
      [201003, 'AdGroupAd'] => 'cm/v201003/',
      [201003, 'AdGroupCriterion'] => 'cm/v201003/',
      [201003, 'AdGroup'] => 'cm/v201003/',
      [201003, 'AdParam'] => 'cm/v201003/',
      [201003, 'BidLandscape'] => 'cm/v201003/',
      [201003, 'BulkMutateJob'] => 'job/v201003/',
      [201003, 'CampaignAdExtension'] => 'cm/v201003/',
      [201003, 'CampaignCriterion'] => 'cm/v201003/',
      [201003, 'Campaign'] => 'cm/v201003/',
      [201003, 'CampaignTarget'] => 'cm/v201003/',
      [201003, 'GeoLocation'] => 'cm/v201003/',
      [201003, 'Info'] => 'info/v201003/',
      [201003, 'Media'] => 'cm/v201003/',
      [201003, 'ReportDefinition'] => 'cm/v201003/',
      [201003, 'TargetingIdea'] => 'o/v201003/',
      # v201008
      [201008, 'AdExtensionOverride'] => 'cm/v201008/',
      [201008, 'AdGroupAd'] => 'cm/v201008/',
      [201008, 'AdGroupCriterion'] => 'cm/v201008/',
      [201008, 'AdGroup'] => 'cm/v201008/',
      [201008, 'AdParam'] => 'cm/v201008/',
      [201008, 'Alert'] => 'mcm/v201008/',
      [201008, 'BidLandscape'] => 'cm/v201008/',
      [201008, 'BulkMutateJob'] => 'job/v201008/',
      [201008, 'CampaignAdExtension'] => 'cm/v201008/',
      [201008, 'CampaignCriterion'] => 'cm/v201008/',
      [201008, 'Campaign'] => 'cm/v201008/',
      [201008, 'CampaignTarget'] => 'cm/v201008/',
      [201008, 'CustomerSync'] => 'ch/v201008/',
      [201008, 'Experiment'] => 'cm/v201008/',
      [201008, 'GeoLocation'] => 'cm/v201008/',
      [201008, 'Info'] => 'info/v201008/',
      [201008, 'Media'] => 'cm/v201008/',
      [201008, 'ReportDefinition'] => 'cm/v201008/',
      [201008, 'ServicedAccount'] => 'mcm/v201008/',
      [201008, 'TargetingIdea'] => 'o/v201008/',
      [201008, 'TrafficEstimator'] => 'o/v201008/',
      [201008, 'UserList'] => 'cm/v201008/'
    }

    # Configure the auth servers to use for each environment
    # The tuplet should be (hostname, port, use_ssl?)
    @@auth_servers = {
      'PRODUCTION' => ['www.google.com', 443, true],
      'SANDBOX' => ['www.google.com', 443, true]
    }

    public

    # Get the available API versions.
    #
    # Returns:
    # List of versions available (as integers)
    #
    def self.get_versions
      @@services.keys
    end

    # Get the latest API version.
    #
    # Returns:
    # Latest version (as an integer)
    #
    def self.get_latest_version
      @@services.keys.select { |service| service.is_a? Integer }.max
    end

    # Does the given environment exist and contain the given version?
    #
    # Returns:
    # Boolean indicating whether the given environment exists and contains the
    # given version
    #
    def self.environment_has_version(environment, version)
      return (!@@environments[environment].nil? and
              !@@environments[environment][version].nil?)
    end

    # Does the given version exist and contain the given service?
    #
    # Returns:
    # Boolean indicating whether the given version exists and contains the
    # given service
    #
    def self.version_has_service(version, service)
      return (!@@services[version].nil? and
              @@services[version].include?(service))
    end

    # Get the default API version.
    #
    # Returns:
    # Default version (as an integer)
    #
    def self.get_default_version
      return DEFAULT_VERSION
    end

    # Get the list of service names for a given version
    #
    # Args:
    # - version: the API version (as an integer)
    #
    # Returns:
    # List of names of services (as strings) available for given version
    #
    def self.get_services(version)
      @@services[version]
    end

    # Get the available environments.
    #
    # Returns:
    # List of available environments (as strings)
    #
    def self.get_environments
      @@environments.keys
    end

    # Get the default environment.
    #
    # Returns:
    # Default environment (as a string)
    #
    def self.get_default_environment
      return DEFAULT_ENVIRONMENT
    end

    # Get the endpoint for a service on a given environment and API version.
    #
    # Args:
    # - environment: the service environment to be used (as a string)
    # - version: the API version (as an integer)
    # - service: the name of the API service (as a string)
    #
    # Returns:
    # The endpoint URL (as a string)
    #
    def self.get_endpoint(environment, version, service)
      base = @@environments[environment][version]
      path = @@subdirs[[version, service]]
      return base.to_s + path.to_s if base
      return nil
    end

    # Get the download URL for reports.
    #
    # Args:
    # - environment: the service environment to be used (as a string)
    # - version: the API version (as an integer)
    #
    # Returns:
    # The endpoint URL (as a string)
    #
    def self.get_report_download_url(environment, version)
      base = @@environments[environment][version]
      path = 'reportdownload' if version >= 201003
      return base.to_s + path.to_s if base and path
      return nil
    end

    # Get the subdirectory for a service, for a given API version.
    #
    # Args:
    # - version: the API version (as an integer)
    # - service: the name of the API service (as a string)
    #
    # Returns:
    # The endpoint URL (as a string)
    #
    def self.get_subdir(version, service)
      return @@subdirs[[version, service]]
    end

    # Get the authentication server details for an environment
    #
    # Args:
    # - environment: the service environment to be used (as a string)
    #
    # Returns:
    # Array containing
    # - the hostname for the auth server (as a string)
    # - the port for the auth server (as an integer)
    # - whether to use SSL or not (as a boolean)
    #
    def self.get_auth_server(environment)
      # If we don't have an entry for this environment, we just return the
      # default server (the same one being used for the default environment)
      if @@auth_servers[environment].nil?
        return @@auth_servers[DEFAULT_ENVIRONMENT]
      end

      return @@auth_servers[environment]
    end

    # Add a new environment to the list.
    #
    # Args:
    # - name: the name for the new environment
    # - endpoint_hash: a hash of base endpoint URLs, indexed by version number,
    #   e.g.:
    #    { 13 => 'URL_FOR_v13', 200906 => 'URL_FOR_v200906' }
    #
    def self.add_environment(name, endpoint_hash)
      @@environments[name] = endpoint_hash
    end

    # Perform the loading of the necessary source files for a version
    #
    # Args:
    # - version: the API version (as an integer)
    #
    def self.do_require(version)
      get_services(version).each do |service|
        eval("require 'adwords4r/v#{version}/#{service}ServiceWrapper.rb'")
      end
    end

    # Returns the full module name for a given service
    #
    # Args:
    # - version: the API version (as an integer)
    # - service: the service name (as a string)
    #
    # Returns:
    # The full module name for the given service (as a string)
    #
    def self.get_module_name(version, service)
      return "AdWords::V#{version.to_s}::#{service}Service"
    end

    # Returns the full interface class name for a given service
    #
    # Args:
    # - version: the API version (as an integer)
    # - service: the service name (as a string)
    #
    # Returns:
    # The full interface class name for the given service (as a string)
    #
    def self.get_interface_name(version, service)
      if version.is_a? Integer and version <= 13
        return get_module_name(version, service) + "::#{service}Interface"
      else
        return get_module_name(version, service) +
            "::#{service}ServiceInterface"
      end
    end

    # Returns the full wrapper class name for a given service
    #
    # Args:
    # - version: the API version (as an integer)
    # - service: the service name (as a string)
    #
    # Returns:
    # The full wrapper class name for the given service (as a string)
    #
    def self.get_wrapper_name(version, service)
      return get_module_name(version, service) + "::#{service}ServiceWrapper"
    end

    # Returns the namespace for a version and service
    #
    # Args:
    # - driver: the service driver
    #
    # Returns:
    # String with the namespace
    def self.get_namespace_v2009(driver)
      # This is a huge hack to get the correct namespace for a given service,
      # but short from configuring it ourselves there seems to be no other
      # option. This accesses one of the fields in the description of the
      # service's methods, which indicates the namespace.
      # Thankfully, since it's a constant and checked as part of our unit tests,
      # it should always work.
      return driver.class::Methods[0][2][0][2][1]
    end

    # Validates whether a parameter is of the correct type
    # This method is invoked by the wrappers during runtime to check the
    # validity of every parameter.
    #
    # Args:
    # - param_name: the parameter name (as a String)
    # - param: the parameter value
    # - type: the expected type (the class object itself)
    #
    # Returns:
    # nil, upon success
    #
    # Raises:
    # - ArgumentError: in case of an unexpected type
    #
    def self.validate_param(param_name, param, type)
      return nil if param.is_a? type

      begin
        type.new(param)
      rescue
        raise ArgumentError, "Parameter '#{param_name}'" +
            " should be convertible into type #{type.to_s}"
      end
      return nil
    end

    # Create a bunch of aliases for the static methods
    class << Service
      alias versions get_versions
      alias latest_version get_latest_version
      alias default_version get_default_version
      alias services get_services
      alias environments get_environments
      alias default_environment get_default_environment
      alias endpoint get_endpoint
      alias report_download_url get_report_download_url
      alias subdir get_subdir
      alias auth_server get_auth_server
      alias module_name get_module_name
      alias interface_name get_interface_name
      alias wrapper_name get_wrapper_name
      alias namespace_v2009 get_namespace_v2009
    end

  end
end
